firstvoices/backend/serializers/widget_serializers.py

```diff
@@ -22,21 +22,31 @@ class Meta:
 
 
 class WidgetDetailSerializer(serializers.ModelSerializer):
-    url = serializers.HyperlinkedIdentityField(view_name="api:widget-detail")
-    type = serializers.CharField(source="widget_type")
-    format = serializers.CharField(source="get_format_display")
-    settings = WidgetSettingsSerializer(source="widgetsettings_set", many=True)
+    url = serializers.HyperlinkedIdentityField(
+        view_name="api:widget-detail", read_only=True
+    )
+    type = serializers.CharField(source="widget_type", required=True)
+    format = serializers.CharField(source="get_format_display", required=True)
+    settings = WidgetSettingsSerializer(
+        source="widgetsettings_set", many=True, required=False
+    )
 
     class Meta:
         model = Widget
         fields = ("url", "id", "title", "type", "format", "settings")
+        read_only_fields = ("id",)
 
 
 class SiteWidgetDetailSerializer(
-    SiteContentLinkedTitleSerializer, WidgetDetailSerializer
+    UpdateSerializerMixin,
+    CreateSiteContentSerializerMixin,
+    SiteContentLinkedTitleSerializer,
+    WidgetDetailSerializer,
 ):
-    url = SiteHyperlinkedIdentityField(view_name="api:sitewidget-detail")
-    visibility = serializers.CharField(source="get_visibility_display")
+    url = SiteHyperlinkedIdentityField(
+        view_name="api:sitewidget-detail", read_only=True
+    )
+    visibility = serializers.CharField(source="get_visibility_display", read_only=True)
```

Comment:
I'm not sure this needs any special permission handling... Based on the chart in our wiki, I thought Language Admins and Superadmins had full access to edit widgets, and nobody else had any edit access.

Reply:
As discussed I have added the ability to edit visibility in this PR.